Full aws-sdk v3 support for nodejs18.x - remove v2 leftovers

## Expected Behavior

No aws-sdk v2 dependencies

## Actual Behavior

Still hidden (as in: not obvious from package.json dependencies) aws-sdk v2 dependencies - e.g. [kms-service](https://github.com/DataDog/datadog-lambda-js/blob/main/src/metrics/kms-service.ts) depends on `aws-sdk/clients/kms`. The source code comment and rationale
```
// In order to avoid the layer adding the 40mb aws-sdk to a deployment, (which is always available
// in the Lambda environment anyway), we use require to import the SDK.
```
is since nodejs18.x not correct anymore. So basically, if I want to use this library in a lambda with nodejs18.x runtime, I need to bundle in the whole aws-sdk-v2 code into the lambda code. 

## Specifications

  - Datadog Lambda Layer version: 7.9.x
  - Node version: 18

## Stacktrace

  ```
 ../app/node_modules/datadog-lambda-js/dist/metrics/kms-service.js:54:38:
      54 │                         kms = require("aws-sdk/clients/kms");
         ╵                                       ~~~~~~~~~~~~~~~~~~~~~

  ```

Any update on this? 
I have node 18 lambda, using aws sdk 3, but cannot run as datadog latest layer still using aws-sdk 2:
`Runtime.ImportModuleError: Error: Cannot find module 'aws-sdk'`

--- a/package.json
+++ b/package.json
@@ -17,6 +17,7 @@
     "postbuild": "scripts/update_dist_version.sh"
   },
   "devDependencies": {
+    "@aws-sdk/client-kms": "^3.366.0",
     "@types/aws-lambda": "^8.10.76",
     "@types/aws-sdk": "^2.7.0",
     "@types/jest": "^26.0.23",

--- a/src/metrics/kms-service.spec.ts
+++ b/src/metrics/kms-service.spec.ts
@@ -39,7 +39,7 @@ describe("KMSService", () => {
       })
       .reply(200, {
         KeyId: KEY_ID,
-        Plaintext: Buffer.from(EXPECTED_RESULT),
+        Plaintext: Buffer.from(EXPECTED_RESULT, "ascii").toString("base64"),
       });
 
     const kmsService = new KMSService();
@@ -83,6 +83,41 @@ describe("KMSService", () => {
         expect((e as Error).message).toEqual(EXPECTED_ERROR_MESSAGE);
       }
     }
+    fakeKmsCall.done();
+  });
+
+  it("decrypts when the API key was encrypted with an encryption context using aws sdk v3", async () => {
+    const fakeKmsCall = nock("https://kms.us-east-1.amazonaws.com:443", { encodedQueryParams: true })
+      .post("/", {
+        CiphertextBlob: "BQICAHj0djbIQaGrIfSD2gstvRF3h8YGMeEvO5rRHNiuWwSeegEFl57KxNejRg==",
+        EncryptionContext: {
+          LambdaFunctionName: "my-test-function",
+        },
+      })
+      .reply(200, {
+        KeyId: KEY_ID,
+        Plaintext: Buffer.from(EXPECTED_RESULT, "ascii").toString("base64"),
+      });
+
+    const kmsService = new KMSService();
+    const result = await kmsService.decryptV3(Buffer.from(ENCRYPTED_KEY, "base64"));
+    expect(Buffer.from(result).toString("ascii")).toEqual(EXPECTED_RESULT);
+    fakeKmsCall.done();
+  });
+
+  it("decrypts when the API key was encrypted without an encryption context using aws sdk v3", async () => {
+    const fakeKmsCall = nock("https://kms.us-east-1.amazonaws.com:443", { encodedQueryParams: true })
+      .post("/", {
+        CiphertextBlob: "BQICAHj0djbIQaGrIfSD2gstvRF3h8YGMeEvO5rRHNiuWwSeegEFl57KxNejRg==",
+      })
+      .reply(200, {
+        KeyId: KEY_ID,
+        Plaintext: Buffer.from(EXPECTED_RESULT, "ascii").toString("base64"),
+      });
+
+    const kmsService = new KMSService();
+    const result = await kmsService.decryptV3(Buffer.from(ENCRYPTED_KEY, "base64"));
+    expect(Buffer.from(result).toString("ascii")).toEqual(EXPECTED_RESULT);
     fakeKmsCall.done();
   });
 });

--- a/src/metrics/kms-service.ts
+++ b/src/metrics/kms-service.ts
@@ -1,14 +1,19 @@
 // In order to avoid the layer adding the 40mb aws-sdk to a deployment, (which is always available
 // in the Lambda environment anyway), we use require to import the SDK.
 
-import { logError } from "../utils";
-
 export class KMSService {
+  private encryptionContext
+
+  constructor () {
+    this.encryptionContext = { LambdaFunctionName: process.env.AWS_LAMBDA_FUNCTION_NAME ?? "" };
+  }
+
   public async decrypt(ciphertext: string): Promise<string> {
+    const buffer = Buffer.from(ciphertext, "base64");
+
     try {
       const kms = require("aws-sdk/clients/kms");
       const kmsClient = new kms();
-      const buffer = Buffer.from(ciphertext, "base64");
 
       // When the API key is encrypted using the AWS console, the function name is added as an encryption context.
       // When the API key is encrypted using the AWS CLI, no encryption context is added.
@@ -19,8 +24,7 @@ export class KMSService {
         result = await kmsClient.decrypt({ CiphertextBlob: buffer }).promise();
       } catch {
         // Then try with encryption context, in case API key was encrypted using the AWS Console
-        const encryptionContext = { LambdaFunctionName: process.env.AWS_LAMBDA_FUNCTION_NAME ?? "" };
-        result = await kmsClient.decrypt({ CiphertextBlob: buffer, EncryptionContext: encryptionContext }).promise();
+        result = await kmsClient.decrypt({ CiphertextBlob: buffer, EncryptionContext: this.encryptionContext }).promise();
       }
 
       if (result.Plaintext === undefined) {
@@ -29,11 +33,26 @@ export class KMSService {
       return result.Plaintext.toString("ascii");
     } catch (err) {
       if ((err as any).code === "MODULE_NOT_FOUND") {
-        const errorMsg = "optional dependency aws-sdk not installed. KMS key decryption will not work";
-        logError(errorMsg);
-        throw Error(errorMsg);
+        // Node 18
+        return this.decryptV3(buffer);
       }
       throw Error("Couldn't decrypt ciphertext");
     }
+  }
+
+  // Node 18 or AWS SDK V3
+  public async decryptV3(buffer: Buffer): Promise<string> {
+    const { KMSClient, DecryptCommand } = require("@aws-sdk/client-kms")
+    const kmsClient = new KMSClient()
+    let result;
+    try {
+      const decryptCommand = new DecryptCommand({ CiphertextBlob: buffer })
+      result = await kmsClient.send(decryptCommand)
+    } catch {
+      const decryptCommand = new DecryptCommand({ CiphertextBlob: buffer, EncryptionContext: this.encryptionContext })
+      result = await kmsClient.send(decryptCommand)
+    }
+    const finalRes = Buffer.from(result.Plaintext).toString("ascii")
+    return finalRes
   }
 }

--- a/src/metrics/kms-service.ts
+++ b/src/metrics/kms-service.ts
@@ -2,9 +2,9 @@
 // in the Lambda environment anyway), we use require to import the SDK.
 
 export class KMSService {
-  private encryptionContext
+  private encryptionContext;
 
-  constructor () {
+  constructor() {
     this.encryptionContext = { LambdaFunctionName: process.env.AWS_LAMBDA_FUNCTION_NAME ?? "" };
   }
 
@@ -24,7 +24,9 @@ export class KMSService {
         result = await kmsClient.decrypt({ CiphertextBlob: buffer }).promise();
       } catch {
         // Then try with encryption context, in case API key was encrypted using the AWS Console
-        result = await kmsClient.decrypt({ CiphertextBlob: buffer, EncryptionContext: this.encryptionContext }).promise();
+        result = await kmsClient
+          .decrypt({ CiphertextBlob: buffer, EncryptionContext: this.encryptionContext })
+          .promise();
       }
 
       if (result.Plaintext === undefined) {
@@ -42,17 +44,17 @@ export class KMSService {
 
   // Node 18 or AWS SDK V3
   public async decryptV3(buffer: Buffer): Promise<string> {
-    const { KMSClient, DecryptCommand } = require("@aws-sdk/client-kms")
-    const kmsClient = new KMSClient()
+    const { KMSClient, DecryptCommand } = require("@aws-sdk/client-kms");
+    const kmsClient = new KMSClient();
     let result;
     try {
-      const decryptCommand = new DecryptCommand({ CiphertextBlob: buffer })
-      result = await kmsClient.send(decryptCommand)
+      const decryptCommand = new DecryptCommand({ CiphertextBlob: buffer });
+      result = await kmsClient.send(decryptCommand);
     } catch {
-      const decryptCommand = new DecryptCommand({ CiphertextBlob: buffer, EncryptionContext: this.encryptionContext })
-      result = await kmsClient.send(decryptCommand)
+      const decryptCommand = new DecryptCommand({ CiphertextBlob: buffer, EncryptionContext: this.encryptionContext });
+      result = await kmsClient.send(decryptCommand);
     }
-    const finalRes = Buffer.from(result.Plaintext).toString("ascii")
-    return finalRes
+    const finalRes = Buffer.from(result.Plaintext).toString("ascii");
+    return finalRes;
   }
 }

--- a/src/metrics/listener.ts
+++ b/src/metrics/listener.ts
@@ -181,9 +181,7 @@ export class MetricsListener {
       try {
         return await this.kmsClient.decrypt(config.apiKeyKMS);
       } catch (error) {
-        if (error instanceof Error) {
-          logError("couldn't decrypt kms api key", error as Error);
-        }
+        logError("couldn't decrypt kms api key", error as Error);
       }
     }
     return "";

